### ISSUE TITLE
Failures on [vm, profiler] Don't assume the isolate has a mutator during a profile sample.

### ISSUE THREAD

**Issue by aam** [dart-lang/sdk#53672]
There are new test failures on [[vm, profiler] Don't assume the isolate has a mutator during a profile sample.](https://github.com/dart-lang/sdk/commit/f36c1094710bd51f643fb4bc84d5de4bfc5d11f3).

[logs](https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8768325255117922801/+/u/test_results/new_test_failures__logs_)

The tests
```
service/awaiter_async_stack_contents_2_test/dds Timeout (expected Pass)
service/awaiter_async_stack_contents_2_test/service Timeout (expected Pass)
service/causal_async_star_stack_contents_test/service Timeout (expected Pass)
service/coverage_optimized_function_test/service Timeout (expected Pass)
service/custom_stream_listen_test/dds Timeout (expected Pass)
service/file_service_test/dds Timeout (expected Pass)
service/file_service_test/service Timeout (expected Pass)
service/get_allocation_traces_test/dds Timeout (expected Pass)
service/get_allocation_traces_test/service Timeout (expected Pass)
service/get_heap_map_rpc_test/dds Timeout (expected Pass)
service/get_instances_rpc_test/dds Timeout (expected Pass)
service/get_object_rpc_test/service Timeout (expected Pass)
service/get_source_report_const_coverage_test/dds Timeout (expected Pass)
service/get_source_report_const_coverage_test/service Timeout (expected Pass)
service/instance_field_order_rpc_test/dds Timeout (expected Pass)
service/instance_field_order_rpc_test/service Timeout (expected Pass)
service/next_through_for_loop_with_break_and_continue_test/dds Timeout (expected Pass)
service/next_through_for_loop_with_break_and_continue_test/service Timeout (expected Pass)
```
are failing on configurations
```
cfe-strong-linux
```

cc @rmacnak-google

**Comment by rmacnak-google** [comment]
The [two](https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8768573932988501345/+/u/test_results/new_test_failures__logs_) [previous](https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8768586693192695201/+/u/test_results/new_test_failures__logs_) failures for this builder show a similar pattern for unrelated commits.

**Comment by aam** [comment]
Turned green, red again https://ci.chromium.org/ui/p/dart/builders/ci.sandbox/front-end-nnbd-linux-release-x64/14445/overview

**Comment by rmacnak-google** [comment]
When I run locally, the longest test runs in under 4 seconds. The timeout is 1 minute, so it's more likely something is stuck than slow.

**Comment by johnniwinther** [comment]
@rmacnak-google Why do you consider these as front-end related? Isn't the slowness a consequence of the tests themselves?

**Comment by rmacnak-google** [comment]
This is on front-end-nnbd-linux-release-x64, where the tests are compiled but not executed.

**Comment by johnniwinther** [comment]
Ah, yes. That might indicate that it has to with assert or verification that we perform on these but, given that we can compile in time when part of the VM.

**Comment by johnniwinther** [comment]
@jensjoha Can you take a look at this? I don't seem to be able to replicate the try-bot run on my Windows machine.

**Comment by jensjoha** [comment]
The TL;DR here is that these tests are all run on a batch compiler, and that the first test pays the cost of the batch compiler starting up. https://dart-review.googlesource.com/c/sdk/+/332220 adds some more timeout for the first run to offset this, which should make the timeouts go away.

Details:

These tests are all run on a (`-j` defaulting to the number of cores) batch compiler(s).
On my machine launching *one* batch compiler (from source) takes 6-7 seconds. Launching 12 (my number of cores) takes 12+ seconds instead. Likely this is more on the bots.
The cost of the startup is attributed to the first test being run. So if the batch compiler takes 59 seconds to start (hopefully it doesn't but for the sake of argument) there's only 1 second left for the test to compile before we get a timeout.
This - for the service tests in particular - is further worsened by these compiling something like 470 libraries (it pulls in package:dds, package:observatory, package:path, package:test, package:yaml etc etc etc.) (whereas many language tests only compile 1 library).
This is in turn even further worsened by the first compile taking longer as the JIT hasn't warmed up yet. When running 12 simultanions compiles (again, the default) I get actual compile times like this:
Compile 1: 22-23 seconds
Compile 2: 9-10 seconds
Compile 3: 6-7 seconds
Compile 4: 5.5-6.5 seconds

Meaning that on my machine the first test takes ~36 seconds. While that leaves a good amount of room before hitting the timeout of 60 seconds, the bots are likely slower and boom.

In https://dart-review.googlesource.com/c/sdk/+/332220 I give the first test run 30 additional seconds of timeout which should fix the timeout.

Generally we'd likely want faster batch compiler startup, which should be possible with AOT compilation of the batch compiler, but that doesn't support asserts (https://github.com/dart-lang/sdk/issues/53343) which is currently enabled and that we probably *do* want to have enabled.

I used `tools/test.py '--progress=status' --report --time --write-results --write-logs --clean-exit -ncfe-strong-linux service/next_...` for this btw.

**Comment by johnniwinther** [comment]
@whesse I think this might related to what you saw recently.
